core/src/main/scala/eu/shiftforward/apso/akka/http/ProxySupport.scala

```diff
@@ -54,47 +55,75 @@ trait ProxySupport extends ClientIPDirectives {
 
-  /**
-   * Proxies a single request to a destination URI.
-   *
-   * @param uri the target URI
-   * @return a route that handles requests by proxying them to the given URI.
-   */
-  def proxySingleTo(uri: Uri): Route = {
+  private[this] def proxy(strictTimeout: Option[FiniteDuration] = None)(reqBuilder: (Option[RemoteAddress], RequestContext) => HttpRequest): Route = {
     extractActorSystem { implicit system =>
       extractMaterializer { implicit mat =>
         optionalRemoteAddress { ip => ctx =>
-          val req = ctx.request.copy(
-            uri = uri,
-            headers = getHeaders(ip, ctx.request.headers.toList))
-
+          val req = reqBuilder(ip, ctx)
           import system.dispatcher
-          Http(system).singleRequest(req).map(Complete.apply)
+          strictTimeout match {
+            case None => Http(system).singleRequest(req).map(Complete.apply)
+            case Some(t) => Http(system).singleRequest(req)
+              .flatMap(r => r.entity.toStrict(t).map(e => r.withEntity(e)))
+              .map(Complete.apply)
+          }
         }
       }
     }
   }
 
+  /**
+   * Proxies a single request to a destination URI.
+   *
+   * @param uri the target URI
+   * @return a route that handles requests by proxying them to the given URI.
+   */
+  def proxySingleTo(uri: Uri): Route = proxy() {
+    case (ip, ctx) => ctx.request.copy(
+      uri = uri,
+      headers = getHeaders(ip, ctx.request.headers.toList))
+  }
+
   /**
    * Proxies a single request to a destination base URI. The target URI is created by concatenating the base URI with
    * the unmatched path.
    *
    * @param uri the target base URI
    * @return a route that handles requests by proxying them to the given URI.
    */
-  def proxySingleToUnmatchedPath(uri: Uri): Route = {
-    extractActorSystem { implicit system =>
-      extractMaterializer { implicit mat =>
-        optionalRemoteAddress { ip => ctx =>
-          val req = ctx.request.copy(
-            uri = uri.withPath(uri.path ++ ctx.unmatchedPath).withQuery(ctx.request.uri.query()),
-            headers = getHeaders(ip, ctx.request.headers.toList))
+  def proxySingleToUnmatchedPath(uri: Uri): Route = proxy() {
+    case (ip, ctx) => ctx.request.copy(
+      uri = uri.withPath(uri.path ++ ctx.unmatchedPath).withQuery(ctx.request.uri.query()),
+      headers = getHeaders(ip, ctx.request.headers.toList))
+  }
 
-          import system.dispatcher
-          Http(system).singleRequest(req).map(Complete.apply)
-        }
-      }
-    }
+  /**
+   * Proxies a single request to a destination URI.
+   * The response in not streamed, but converted to a strict entity with a set timeout.
+   *
+   * @param uri the target URI
+   * @param timeout maximum time to wait for the full response.
+   * @return a route that handles requests by proxying them to the given URI.
+   */
+  def proxySingleToStrict(uri: Uri, timeout: FiniteDuration): Route = proxy(Some(timeout)) {
```

Comment:
Nitpick: don't you prefer `strictProxySingleTo`? The `ToStrict` at the end seems strange.